tuf/client_rework/metadata_bundle.py

```diff
@@ -0,0 +1,505 @@
+            if unverified.verify(key):
+                unique_keys.add(key["keyval"]["public"])
+        except Exception as e:  # pylint: disable=broad-except
+            # TODO specify the Exceptions (see issue #1351)
+            logger.info("verify failed: %s", e)
+
+    return len(unique_keys) >= role["threshold"]
+
+
+class MetadataBundle(abc.Mapping):
+    """Internal class to keep track of valid metadata in Updater
+
+    MetadataBundle ensures that metadata is valid. It provides easy ways to
+    update the metadata with the caller making decisions on what is updated.
+    """
+
+    def __init__(self, repository_path: str):
+        """Initialize by loading root metadata from disk"""
+        self._path = repository_path
+        self._bundle = {}  # type: Dict[str: Metadata]
+        self.reference_time = None
+
+        # Load and validate the local root metadata
+        # Valid root metadata is required
+        logger.debug("Loading local root")
+        try:
+            with open(os.path.join(self._path, "root.json"), "rb") as f:
+                self._load_intermediate_root(f.read())
+        except (OSError, exceptions.RepositoryError) as e:
+            raise exceptions.RepositoryError(
+                "Failed to load local root metadata"
+            ) from e
+
+    # Implement Mapping
+    def __getitem__(self, key: str) -> Metadata:
+        return self._bundle[key]
+
+    def __len__(self) -> int:
+        return len(self._bundle)
+
+    def __iter__(self) -> Iterator[Metadata]:
+        return iter(self._bundle)
+
+    # Helper properties for top level metadata
+    @property
+    def root(self) -> Optional[Metadata]:
+        return self._bundle.get("root")
+
+    @property
+    def timestamp(self) -> Optional[Metadata]:
+        return self._bundle.get("timestamp")
+
+    @property
+    def snapshot(self) -> Optional[Metadata]:
+        return self._bundle.get("snapshot")
+
+    @property
+    def targets(self) -> Optional[Metadata]:
+        return self._bundle.get("targets")
+
+    # Public methods
+    def update_root(self, data: bytes):
+        """Update root metadata with data from remote repository."""
+        logger.debug("Updating root")
+
+        self._load_intermediate_root(data)
+        self.root.to_file(os.path.join(self._path, "root.json"))
+
+    def root_update_finished(self):
+        """Mark root metadata as final."""
+        if self.reference_time is not None:
+            raise ValueError("Root update is already finished")
+
+        # Store our reference "now", verify root expiry
+        self.reference_time = datetime.utcnow()
+        if self.root.signed.is_expired(self.reference_time):
+            raise exceptions.ExpiredMetadataError("New root.json is expired")
+
+        logger.debug("Verified final root.json")
+
+    def load_local_timestamp(self) -> bool:
+        """Load cached timestamp metadata from local storage.
+
+        Returns True if timestamp was succesfully loaded"""
```

Comment:
This was a surprise. Especially as there were not any signs of checking return type of the `load_local_*` functions in the example snippet near the top of the file. Wouldn't the Pythonic way to do this be to throw an exception if the `load_local_*()` functions do not succeed? Or was this intentional, because only root metadata **must** exist locally?

Reply:
Exceptions work fin, I'll use them throughout